src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs

```diff
@@ -0,0 +1,352 @@
+                }
+            }
+            else
+            {
+                var dic = GetContractDictionary(serviceType, false);
+                if (dic != null && dic.TryRemove(contract, out var _) && dic.Count == 0)
+                {
+                    RemoveContractService(serviceType);
+                }
+            }
+
+            // required so that it gets rebuilt if not injected externally.
+            _serviceProvider = null;
+        }
+
+        /// <inheritdoc />
+        public virtual IDisposable ServiceRegistrationCallback(Type serviceType, string contract, Action<IDisposable> callback)
+        {
+            // this method is not used by RxUI
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public virtual bool HasRegistration(Type serviceType)
+        {
+            if (serviceType == null)
+            {
+                serviceType = typeof(NullServiceType);
+            }
+
+            if (!_isImmutable)
+            {
+                return _serviceCollection.Any(sd => sd.ServiceType == serviceType);
+            }
+
+            var service = _serviceProvider.GetService(serviceType);
+            return service != null;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes of the instance.
+        /// </summary>
+        /// <param name="disposing">Whether or not the instance is disposing.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        private static Type GetDictionaryType(Type serviceType) => _dictionaryType.MakeGenericType(serviceType);
+
+        private void RemoveContractService(Type serviceType)
+        {
+            var dicType = GetDictionaryType(serviceType);
+            var sd = _serviceCollection.SingleOrDefault(s => s.ServiceType == serviceType);
+            _serviceCollection.Remove(sd);
+        }
+
+        private ContractDictionary GetContractDictionary(Type serviceType, bool createIfNotExists)
+        {
+            var dicType = GetDictionaryType(serviceType);
+
+            if (_isImmutable)
+            {
+                return (ContractDictionary)ServiceProvider.GetService(dicType);
+            }
+
+            var dic = _serviceCollection
+                .Where(sd => sd.ServiceType == dicType)
+                .Select(sd => sd.ImplementationInstance)
+                .Cast<ContractDictionary>()
+                .SingleOrDefault();
+
+            if (createIfNotExists && dic == null)
+            {
+                dic = (ContractDictionary)Activator.CreateInstance(dicType);
+                _serviceCollection.AddSingleton(dicType, dic);
+            }
+
+            return dic;
+        }
+
+        private class ContractDictionary : ConcurrentDictionary<string, IList<Func<object>>>
```

Comment:
Don't inherit from concurrent dictionary. Use composition.

Also use the atomic operators such as GetOrAdd with the func override etc.

Reply:
> Use composition.

Why?

> Also use the atomic operators

Right, ok.

Reply:
Done.